### **DSB6_MovieLens Analytics** 
#### Team Lead<br>
- *Anna* (janaspic)<br>

#### Team Members<br>
- *Anna* (janaspic)<br>
- *Anastasia* (berenhak)<br>

 ### **ML-Latest-Small data description**
<hr>

#### `ratings.csv` - оценки пользователей
- **userId** - id пользователя  
- **movieId** - id фильма  
- **rating** - оценка фильма  
- **timestamp** - время оценки  

---

#### `movies.csv` - фильмы и жанры
- **movieId** - id фильма  
- **title** - название фильма  
- **genres** - жанры фильма  

---

#### `tags.csv` - пользовательские теги
- **userId** - id пользователя  
- **movieId** - id фильма  
- **tag** - текст тега  
- **timestamp** - время добавления тега  

---

#### `links.csv` - внешние ссылки
- **movieId** - id фильма  
- **imdbId** - id в IMDb  
- **tmdbId** - id в TMDb  

In [3]:
%timeit
import movielens_analysis as ma

print("Модуль movielens_analysis успешно подключён!")

Модуль movielens_analysis успешно подключён!


In [4]:
%timeit
from ratings import average, median

print("Загружены необходимые методы!")

Загружены необходимые методы!


### **Анализ movies.csv**

In [13]:
%timeit
dM=ma.Movies(limit=1000)
dM.get_movies()[:2]

[{'movieID': '1',
  'title': 'Toy Story',
  'release': '1995',
  'genres': ['Adventure', 'Animation', 'Children', 'Comedy', 'Fantasy']},
 {'movieID': '2',
  'title': 'Jumanji',
  'release': '1995',
  'genres': ['Adventure', 'Children', 'Fantasy']}]

>Возвращает список словарей с информацией о фильмах: {ID, название, жанры, год}

In [4]:
%timeit
#Распределение фильмов по годам выпуска
dM.dist_by_release()

{'1995': 224,
 '1994': 184,
 '1996': 181,
 '1993': 101,
 '1992': 23,
 '1990': 15,
 '1991': 15,
 '1989': 14,
 '1986': 9,
 '1982': 8,
 '1940': 8,
 '1957': 8,
 '1987': 8,
 '1980': 8,
 '1981': 7,
 '1988': 7,
 '1979': 7,
 '1955': 6,
 '1959': 6,
 '1968': 6,
 '1997': 6,
 '1939': 6,
 '1985': 6,
 '1967': 5,
 '1965': 5,
 '1951': 5,
 '1958': 5,
 '1944': 5,
 '1941': 5,
 '1975': 5,
 '1971': 5,
 '1984': 5,
 '1964': 4,
 '1973': 4,
 '1954': 4,
 '1934': 4,
 '1960': 4,
 '1963': 4,
 '1950': 4,
 '1974': 4,
 '1983': 4,
 '1977': 3,
 '1937': 3,
 '1972': 3,
 '1952': 3,
 '1961': 3,
 '1953': 3,
 '1946': 3,
 '1938': 3,
 '1962': 3,
 '1976': 2,
 '1969': 2,
 '1970': 2,
 '1942': 2,
 '1945': 2,
 '1947': 2,
 '1935': 2,
 '1936': 2,
 '1956': 2,
 '1949': 2,
 '1978': 2,
 '1943': 1,
 '1932': 1,
 '1966': 1,
 '1948': 1,
 '1933': 1,
 '1931': 1}

>Возвращает словарь, где ключом является год, значением - количество фильмов, выпущенных в этот год
<hr>

In [5]:
%timeit
# Распределение фильмов по жанрам
dM.dist_by_genres()

{'Drama': 506,
 'Comedy': 365,
 'Romance': 208,
 'Thriller': 179,
 'Action': 158,
 'Adventure': 126,
 'Crime': 122,
 'Children': 100,
 'Fantasy': 69,
 'Sci-Fi': 68,
 'Mystery': 58,
 'Musical': 53,
 'Horror': 51,
 'War': 48,
 'Animation': 37,
 'Documentary': 25,
 'Western': 23,
 'Film-Noir': 18,
 'IMAX': 3}

>Возвращает словарь, где ключом является жанр, значением - количество фильмов в этом жанре
<hr>

In [6]:
%timeit
# Посмотрим процентное соотношение жанров в нашей выборке
total_genres=sum(dM.dist_by_genres().values())
genres_share={}    

for k,v in dM.dist_by_genres().items():
    genres_share[k]=round(float((v/total_genres)*100), 2)

genres_share

{'Drama': 22.82,
 'Comedy': 16.46,
 'Romance': 9.38,
 'Thriller': 8.07,
 'Action': 7.13,
 'Adventure': 5.68,
 'Crime': 5.5,
 'Children': 4.51,
 'Fantasy': 3.11,
 'Sci-Fi': 3.07,
 'Mystery': 2.62,
 'Musical': 2.39,
 'Horror': 2.3,
 'War': 2.17,
 'Animation': 1.67,
 'Documentary': 1.13,
 'Western': 1.04,
 'Film-Noir': 0.81,
 'IMAX': 0.14}

<hr>

In [7]:
%timeit
# Топ 5 жанров по количеству фильмов
dM.most_genres(5)

{'Drama': 506, 'Comedy': 365, 'Romance': 208, 'Thriller': 179, 'Action': 158}

#### **Вывод по movies.csv**

Проанализировав 1000 фильмов из movies.csv, мы получили следующие инсайты: 
- Большую часть фильмов выпустили с **1992 - 1995 годы**. Это может быть связано как с особенностями формирования датасета, так и с тем, что в начале 90-ых произошёл бум формата видеокассет, что сделало их доступными для более широкого круга аудитории
- Самый популярный жанр - **Драма** (22%), второй по популярности - **Комедия** (16%). Наименьшей популярностью обладают фильмы, снятые в формате **IMAX** (<1%)- поскольку съёмка в данном формате является более дорогостоящей, и в жанре **Нуар** (<1%) - он был наиболее актуален в 1940-50-ые годы, а большую часть выборки составляют фильмы из других десятилетий

### **Анализ tags.csv**

In [9]:
%timeit
dT=ma.Tags(limit=1000)
dT.get_tags()[:3]

[{'userId': 2, 'movieId': 60756, 'tag': 'funny', 'timestamp': 1445714994},
 {'userId': 2,
  'movieId': 60756,
  'tag': 'Highly quotable',
  'timestamp': 1445714996},
 {'userId': 2,
  'movieId': 60756,
  'tag': 'will ferrell',
  'timestamp': 1445714992}]

>Возвращает список словарей с информацией по тегу: {user_id, movie_id, сам тег и время присвоения тега}
<hr>

In [14]:
%timeit
# Топ 5 тегов по количству слов
dT.most_words(5)

{'something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'oscar (best music - original score)': 6,
 'everything you want is here': 5,
 'based on a true story': 5}

>Возвращает словарь заданной длины, ключ - тег, значение - количество слов в нем
<hr>

In [15]:
%timeit
# Топ 5 тегов по количеству символов
dT.longest(5)

{'something for everyone in this one... saw it without and plan on seeing it with kids!': 85,
 'the catholic church is the most corrupt organization in history': 63,
 'audience intelligence underestimated': 36,
 'oscar (best music - original score)': 35,
 'assassin-in-training (scene)': 28}

>Возвращает словарь заданной длины, ключ - тег, значение - количество символов в нем
<hr>

In [16]:
%timeit
# Топ 5 тегов по количеству слов и символов
dT.most_words_and_longest(5)

['oscar (best music - original score)',
 'something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history']

>Возвращает список тегов, которые входят в два предыдущит ТОПа одновременно
<hr>

In [17]:
%timeit
# Самые часто встречающиеся теги 
dT.most_popular(5)

{'funny': 15,
 'sci-fi': 15,
 'twist ending': 13,
 'action': 13,
 'dark comedy': 12}

>Возвращает словарь заданной длины, где ключ - тег, значение - частота встречаемости тега
<hr>

In [ ]:
%timeit
# Теги, содержащие строку 'fun'
dT.tags_with('fun')

['fun family movie', 'funny', 'stupid but funny', 'very funny']

>Возвращает список тегов с совпадениями по данной строке
<hr>

In [19]:
%timeit
# Распределение тегов по годам
dT.dist_by_year()

{2018: 412,
 2016: 355,
 2015: 79,
 2012: 45,
 2006: 41,
 2009: 13,
 2010: 13,
 2013: 9,
 2007: 8,
 2014: 7,
 2008: 7,
 2011: 7,
 2017: 3}

>Возвращает год и количество тегов, опубликованных в этот год
<hr>

#### **Вывод по tags.csv**

Мы проанализиировали первые 1000 тегов из данного файла. Были сделаны следующие выводы: 

- Самый длинный тег содержит целых *16* слов:<br>
> something for everyone in this one... saw it without and plan on seeing it with kids!<br>
- Этот же тег лидирует и по количеству символов, их *85*. <br>
- Самые популярные теги: **funny, sci-fi, twist ending, action, dark comedy**. Это может косвенно указывать на интерес пользователей платформы к определенным жанрам. <br>
- В **2018 и 2016** годах был пик тегирования фильмов, что может говорить о растущей популярности платформы в эти годы.<br>

### **Анализ ratings.csv**

Объект класса Ratings хранит в себе информацию об отзывах пользователей на первую 1000 фильмов из movies.csv

In [10]:
%timeit
dR=ma.Ratings("./ml-latest-small/ratings.csv", "./ml-latest-small/movies.csv")
dR.data[:2]

[{'userId': '1',
  'movieId': '1',
  'title': 'Toy Story (1995)',
  'genres': 'Adventure|Animation|Children|Comedy|Fantasy',
  'rating': 4.0,
  'timestamp': 964982703},
 {'userId': '1',
  'movieId': '3',
  'title': 'Grumpier Old Men (1995)',
  'genres': 'Comedy|Romance',
  'rating': 4.0,
  'timestamp': 964981247}]

>Возвращает список словарей формата {userId, movieId название фильма, его жанры, оценку и время выставления оценки}
<hr>

In [13]:
%timeit
# Обращаемся к внутреннему классу Movies через атрибут объекта Ratings
dR.movies._data[:2]

[{'userId': '1',
  'movieId': '1',
  'title': 'Toy Story (1995)',
  'genres': 'Adventure|Animation|Children|Comedy|Fantasy',
  'rating': 4.0,
  'timestamp': 964982703},
 {'userId': '1',
  'movieId': '3',
  'title': 'Grumpier Old Men (1995)',
  'genres': 'Comedy|Romance',
  'rating': 4.0,
  'timestamp': 964981247}]

In [14]:
%timeit
# Распределение оценок по годам
dR.movies.dist_by_year()

{1996: 6001,
 1997: 1728,
 1998: 276,
 1999: 1105,
 2000: 3788,
 2001: 1119,
 2002: 931,
 2003: 890,
 2004: 881,
 2005: 1591,
 2006: 1062,
 2007: 1610,
 2008: 988,
 2009: 984,
 2010: 512,
 2011: 415,
 2012: 1001,
 2013: 390,
 2014: 206,
 2015: 1137,
 2016: 978,
 2017: 1295,
 2018: 745}

>Возвращает словарь, где ключ - год, значение - количество оценок, проставленных в этом году
<hr>

In [15]:
%timeit
# Распределение по значениям оценок
dR.movies.dist_by_rating()

{0.5: 221,
 1.0: 824,
 1.5: 289,
 2.0: 2035,
 2.5: 958,
 3.0: 6818,
 3.5: 2412,
 4.0: 8402,
 4.5: 1988,
 5.0: 5686}

>Возвращает словарь, где ключ - оценка, значение - частота встречаемости
<hr>

In [17]:
%timeit
# Топ 5 фильмов по количеству оценок
dR.movies.top_by_num_of_ratings(5)

{'Forrest Gump (1994)': 329,
 'Shawshank Redemption, The (1994)': 317,
 'Pulp Fiction (1994)': 307,
 'Silence of the Lambs, The (1991)': 279,
 'Star Wars: Episode IV - A New Hope (1977)': 251}

>Возвращает словарь заданной длины, ключ - название фильма, значение - количество его оценок
<hr>

In [18]:
%timeit
# ТОП 5 фильмов по средней и медианной оценке
dR.movies.top_by_ratings(5, metric=average)

{'Lesson Faust (1994)': 5.0,
 'Lamerica (1994)': 5.0,
 'Entertaining Angels: The Dorothy Day Story (1996)': 5.0,
 'Heidi Fleiss: Hollywood Madam (1995)': 5.0,
 'Awfully Big Adventure, An (1995)': 5.0}

In [19]:
%timeit
dR.movies.top_by_ratings(5, metric=median)

{'Children of the Corn IV: The Gathering (1996)': 5.0,
 'Lesson Faust (1994)': 5.0,
 'Secrets & Lies (1996)': 5.0,
 'Lamerica (1994)': 5.0,
 'Entertaining Angels: The Dorothy Day Story (1996)': 5.0}

>Возвращает словарь заданной длины, ключ - название фильма, значение - расчёт указанной статистической характеристики рейтинга
<hr>

In [20]:
%timeit
# ТОП 5 самых противоречивых фильмов (по дисперсии)
dR.movies.top_controversial(5)

{'Lassie (1994)': 4.0,
 'Mr. Wrong (1996)': 3.04,
 'Ed (1996)': 2.74,
 'Burnt by the Sun (Utomlyonnye solntsem) (1994)': 2.72,
 'Relic, The (1997)': 2.5}

>Возвращает словарь заданной длины, ключ - название фильма, значение - дисперсия рейтингов
<hr>

In [23]:
%timeit
# Пользователи, оценившие фильм
dR.movies.users_who_rated('My Fair Lady (1964)')

['3',
 '4',
 '20',
 '29',
 '33',
 '50',
 '84',
 '100',
 '104',
 '105',
 '132',
 '140',
 '169',
 '177',
 '186',
 '188',
 '292',
 '315',
 '328',
 '357',
 '398',
 '474',
 '483',
 '488',
 '489',
 '517',
 '525',
 '552',
 '554',
 '563',
 '590',
 '593',
 '600',
 '603',
 '606']

>Возвращает список пользователей
<hr>

In [25]:
%timeit
# Распределение среднего значения оценок фильма по годам
dR.movies.rating_trend_by_year('My Fair Lady (1964)')

{1997: 4.0,
 1999: 4.2,
 2000: 5.0,
 2002: 4.0,
 2003: 4.5,
 2004: 4.25,
 2005: 4.0,
 2006: 3.0,
 2007: 4.5,
 2008: 4.0,
 2009: 4.25,
 2011: 2.75,
 2012: 4.5,
 2013: 3.0,
 2015: 4.0,
 2016: 4.0,
 2017: 4.25,
 2018: 3.0}

>Возвращает словарь, ключ - год, значение - средняя оценка в году
<hr>

In [27]:
%timeit
# ТОП 5 пользователей по активности проставления оценок
dR.users.top_by_num_of_ratings(5)

{'414': 490, '599': 459, '474': 392, '603': 331, '6': 314}

>Возвращает словарь заданной длины, ключ - user_id, значение - количество оценок, проставленных пользователем
<hr>

In [28]:
%timeit
# ТОП 5 пользователей по среднему,медианному значению вытавляемых оценок
dR.users.top_by_ratings(5, metric = average)

{'53': 5.0, '77': 5.0, '106': 5.0, '154': 5.0, '236': 5.0}

In [29]:
%timeit
dR.users.top_by_ratings(5, metric = median)

{'1': 5.0, '25': 5.0, '30': 5.0, '43': 5.0, '52': 5.0}

>Возвращает словарь заданной длины, ключ - user_id, значение - среднее/медианное значение его оценок
<hr>

In [30]:
%timeit
# ТОП 5 пользователей с наибольшей дисперсией выставленных оценок
dR.users.top_controversial(5)

{'393': 3.89, '255': 3.24, '499': 3.1, '55': 3.06, '461': 2.83}

>Возвращает словарь заданной длины, ключ - user_id, значение - дисперсия оценок
<hr>

In [5]:
%timeit
# Число оценок, выставленных пользователями
dR.users.dist_by_num_of_ratings()

{'106': 1,
 '291': 1,
 '320': 1,
 '459': 1,
 '556': 1,
 '2': 2,
 '148': 2,
 '154': 2,
 '175': 2,
 '236': 2,
 '358': 2,
 '481': 2,
 '506': 2,
 '519': 2,
 '578': 2,
 '598': 2,
 '139': 3,
 '258': 3,
 '272': 3,
 '537': 3,
 '545': 3,
 '550': 3,
 '575': 3,
 '581': 3,
 '582': 3,
 '25': 4,
 '55': 4,
 '65': 4,
 '77': 4,
 '114': 4,
 '143': 4,
 '158': 4,
 '207': 4,
 '281': 4,
 '324': 4,
 '333': 4,
 '366': 4,
 '466': 4,
 '471': 4,
 '473': 4,
 '511': 4,
 '13': 5,
 '92': 5,
 '97': 5,
 '127': 5,
 '157': 5,
 '189': 5,
 '190': 5,
 '209': 5,
 '252': 5,
 '278': 5,
 '299': 5,
 '300': 5,
 '388': 5,
 '496': 5,
 '499': 5,
 '515': 5,
 '548': 5,
 '48': 6,
 '73': 6,
 '138': 6,
 '205': 6,
 '407': 6,
 '533': 6,
 '49': 7,
 '87': 7,
 '123': 7,
 '146': 7,
 '237': 7,
 '296': 7,
 '406': 7,
 '460': 7,
 '472': 7,
 '487': 7,
 '491': 7,
 '518': 7,
 '10': 8,
 '67': 8,
 '88': 8,
 '101': 8,
 '248': 8,
 '338': 8,
 '341': 8,
 '383': 8,
 '392': 8,
 '397': 8,
 '430': 8,
 '442': 8,
 '564': 8,
 '591': 8,
 '208': 9,
 '218': 9,
 '24

>Возвращает словарь, где ключ - id пользователя, значение - количество выставленных оценок
<br>

In [6]:
%timeit
# Распределение среднего значения выставляемых оценок пользователями
dR.users.dist_by_ratings(metric = average)

{'3': 1.12,
 '442': 1.38,
 '598': 1.5,
 '308': 1.82,
 '259': 1.94,
 '153': 1.98,
 '175': 2.0,
 '388': 2.1,
 '508': 2.23,
 '55': 2.25,
 '207': 2.25,
 '89': 2.26,
 '311': 2.33,
 '576': 2.39,
 '567': 2.4,
 '293': 2.5,
 '230': 2.52,
 '297': 2.6,
 '329': 2.62,
 '428': 2.62,
 '294': 2.63,
 '36': 2.66,
 '19': 2.71,
 '365': 2.71,
 '281': 2.75,
 '386': 2.75,
 '599': 2.75,
 '81': 2.77,
 '217': 2.77,
 '287': 2.78,
 '431': 2.79,
 '608': 2.79,
 '10': 2.81,
 '76': 2.82,
 '517': 2.83,
 '214': 2.84,
 '406': 2.86,
 '307': 2.87,
 '361': 2.88,
 '245': 2.89,
 '478': 2.91,
 '34': 2.92,
 '255': 2.92,
 '316': 2.92,
 '181': 2.94,
 '160': 2.96,
 '394': 2.96,
 '600': 2.96,
 '149': 2.97,
 '41': 2.98,
 '132': 2.98,
 '133': 3.0,
 '139': 3.0,
 '143': 3.0,
 '223': 3.0,
 '466': 3.0,
 '537': 3.0,
 '510': 3.02,
 '54': 3.03,
 '68': 3.03,
 '94': 3.04,
 '395': 3.05,
 '500': 3.05,
 '549': 3.05,
 '314': 3.06,
 '28': 3.07,
 '47': 3.07,
 '238': 3.08,
 '368': 3.1,
 '208': 3.11,
 '262': 3.11,
 '461': 3.11,
 '571': 3.11,
 '158':

>Возвращает словарь, ключ - id пользователя, значение - статистическая характеристика выставленных им оценок
<br>

In [31]:
%timeit
# ТОП 5 активных пользователей за определенный год
dR.users.top_active_by_period(5, 1996)

{'6': 314, '117': 165, '602': 135, '109': 127, '411': 127}

>Возвращает словарь заданной длины, ключ - user_id, значение - количество его отзывов за указанный год
<hr>

In [32]:
%timeit
# Список фильмов, оцененных определенным пользователем
dR.users.movies_rated_by_user(3)

['Dangerous Minds (1995)',
 "Schindler's List (1993)",
 'Courage Under Fire (1996)',
 'Operation Dumbo Drop (1995)',
 'Wallace & Gromit: The Best of Aardman Animation (1996)',
 'Escape from L.A. (1996)',
 'My Fair Lady (1964)',
 'Doors, The (1991)',
 'On Golden Pond (1981)',
 'Deer Hunter, The (1978)',
 'Patton (1970)',
 'Highlander (1986)']

In [34]:
%timeit
# ТОП 5 фильмов с наивысшими оценками пользователя
dR.users.top_movies_for_user(3, 5)

{'Escape from L.A. (1996)': 5.0,
 'Highlander (1986)': 3.5,
 'Dangerous Minds (1995)': 0.5,
 "Schindler's List (1993)": 0.5,
 'Courage Under Fire (1996)': 0.5}

#### **Вывод по ratings.csv**

Проведя анализ отзывов о 1000 фильмах, были сделаны следующие выводы: 

- **Пик активности** пришёлся на период *1996-2000 гг.*, особенно в 1996 году (6001 оценка).

- Наиболее часто встречающиеся оценки:

    - 4.0 - 8402 оценки
    - 3.0 - 6818 оценок
    - 5.0 - 5686 оценок

  **-> Распределение оценок смещено влево, в сторону более высоких оценок, и бимодально, поскольку пользователи не склонны ставить дробную оценку**

- Топ фильмов включает по статистическим характеристикам включает в себя фильмы:
    > Lesson Faust, Lamerica, Heidi Fleiss: Hollywood Madam

- Средняя и медианная оценка у ряда фильмов достигает максимума

- Самыми противоречивыми (по дисперсии) фильмами стали:
    >Lassie (1994): 4.0, <br>
    >Mr. Wrong (1996): 3.04 <br>
- В разрезе пользователей:

    Пользователь 414 - самый активный (490 оценок)<br>
    Пользователь 393 - наиболее противоречивый (дисперсия 3.89)<br>
    Средняя оценка у большинства пользователей - выше 3.5, что **говорит о позитивном восприятии фильмов**.<br>

### **Анализ link.csv**

In [11]:
%timeit
dL=ma.Links("./ml-latest-small/links.csv", "./ml-latest-small/movies.csv")
dL.data[:3]

[{'movieId': '1',
  'imdbId': '0114709',
  'tmdbId': '862',
  'title': 'Toy Story (1995)',
  'genres': 'Adventure|Animation|Children|Comedy|Fantasy'},
 {'movieId': '2',
  'imdbId': '0113497',
  'tmdbId': '8844',
  'title': 'Jumanji (1995)',
  'genres': 'Adventure|Children|Fantasy'},
 {'movieId': '3',
  'imdbId': '0113228',
  'tmdbId': '15602',
  'title': 'Grumpier Old Men (1995)',
  'genres': 'Comedy|Romance'}]

In [13]:
%timeit
# Получить информацию о фильме по imdb_id фильма
dL.get_imdb([1, 2, 3], ["title", "budget"])

[[3, 'Grumpier Old Men', 25000000],
 [2, 'Jumanji', 65000000],
 [1, 'Toy Story', 30000000]]

In [14]:
%timeit
# Получить информацию о фильме по tmdb_id фильма
dL.get_tmdb([1, 2, 3], ["title", "budget"])

[[3, 'Grumpier Old Men', 25000000],
 [2, 'Jumanji', 65000000],
 [1, 'Toy Story', 30000000]]

>Возвращает список списков, содержащий [movie_id, название и бюджет фильма] (доступные поля описаны в документации)
<hr>

Первый запуск любой функции расчёта топов по характеристикам, полученным из imdb, будет занимать около 10 минут, поскольку необходимо получить данные о 1000 фильмах. Но благодаря созданию кэша данные извлекаются 1 раз для каждого фильма, и последующие расчёты будут занимать секунды

In [15]:
%timeit
# Топ 5 самых часто встречающихся режиссёров среди фильмов
dL.top_directors(5)

{'Alfred Hitchcock': 11,
 'Woody Allen': 8,
 'Wilfred Jackson': 6,
 'Stanley Kubrick': 6,
 'Frank Capra': 6}

>Возвращает словарь заданной длины, ключ - режиссёр, значение - количество фильмов

Попробуем запустить функцию ещё раз, чтобы убедиться во времени выполнения при обращении к данным из кэша

In [16]:
%timeit
dL.top_directors(5)

{'Alfred Hitchcock': 11,
 'Woody Allen': 8,
 'Wilfred Jackson': 6,
 'Stanley Kubrick': 6,
 'Frank Capra': 6}

In [17]:
%timeit
# Фильмы с наибольшим бюджетом
dL.most_expensive(5)

{'Akira': 1100000000,
 'Ghost in the Shell': 330000000,
 'Waterworld': 175000000,
 'Germinal': 164000000,
 'Cold Fever': 130000000}

>Возвращает словарь заданной длины, ключ - название фильма, значение - бюджет
<hr>

In [18]:
%timeit
# Фильмы с наибольшей разницей между сборами и бюджетом
dL.most_profitable(5)

{'Jurassic Park': 1040110411,
 'The Lion King': 934161632,
 'E.T. the Extra-Terrestrial': 786807407,
 'Star Wars: Episode IV - A New Hope': 764398507,
 'Independence Day': 742400891}

>Возвращает словарь заданной длины, ключ - название фильма, значение - прибыль 
<hr>

In [19]:
%timeit
# Самые длинные фильмы по хронометражу 
dL.longest(5)

{'Gone with the Wind': 238,
 'Once Upon a Time in America': 229,
 'Lawrence of Arabia': 227,
 'Ben-Hur': 212,
 'The Godfather Part II': 202}

>Возвращает словарь заданной длины, ключ - название фильма, значение - продолжительность в минутах
<hr>

In [20]:
%timeit
# Фильмы с самой дорогой стоимостью одной минуты хронометража
dL.top_cost_per_minute(5)

{'Akira': 8870967.74,
 'Ghost in the Shell': 3975903.61,
 'Cold Fever': 1566265.06,
 'Waterworld': 1296296.3,
 'The Hunchback of Notre Dame': 1098901.1}

>Возвращает словарь заданной длины, ключ - название фильма, значение - стоимость одной минуты
<hr>

#### **Вывод по links.csv**

Мы проанализиировали данный файл и можем сделать следующие выводы:
- Режиссёры:

Самыми популярными режиссерами стали, действительно, культовые режиссеры такие как:
    - Alfred Hitchcock
    - Woody Allen
    - Wilfred Jackson
    - Stanley Kubrick

- Бюджет:

Самый дорогой фильм из анализируемых - с бюджетом 1,1 млрд долларов, далее следуют фильмы с бюджетами до 330 млн долларов.

- Прибыльность:

Наиболее прибыльный фильм заработал 104 млн долларов, однако в выборке также есть фильмы с убытками, что указывает на коммерческие провалы несмотря на затраты.

- Продолжительность:

Самый длинный фильм длится почти 4 часа большинство других - не дольше 229 минут.

- Стоимость одной минуты фильма: 

Стоимость производства одной минуты достигает до 8,9 млн долларов/мин - впечатляющий показатель для высокобюджетных картин!

## Финальный итог анализа
В ходе выполнения проекта мы:

-- Проанализировали 4 основных источника данных:<br>

    *movies.csv* - жанры и годы выпуска 
    *ratings.csv* - оценки пользователей и активность
    *tags.csv* - смысловые теги и поведенческие паттерны
    *links.csv + IMDb* - бюджеты, кассовые сборы, режиссёры и продолжительность

Использовали собственные классы Movies, Ratings, Tags, Links для структурированной и расширяемой обработки данных<br>

-- Получили инсайты:<br>

 * Основной массив фильмов - 1992–1995 гг., с доминированием жанров драма и комедия<br>
 * Пользователи склонны ставить средние и высокие оценки (3.0–5.0)<br>
 * Популярные теги - funny, sci-fi, action<br>
 * Существуют как высокодоходные фильмы, так и провальные по бюджету<br>
 * Стоимость производства одной минуты фильма может достигать миллиона долларов<br>

 Весь анализ помог нам детально изучить данные платформы, углубить свои знания ООП, работы с API и обработки большого количества данных <br>